### ISSUE TITLE
Allow zkg's config knowledge to be available in plugin configure scripts

### ISSUE THREAD

**Issue by philrz** [zeek/package-manager#68]
I [discussed this one on Slack](https://zeekorg.slack.com/archives/CTGDGK6EA/p1595799834095300) with @ckreibich.

In my case, I'd already successfully/previously completed `zkg autoconfig` at a time when `zeek-config` was in my path. Now I was circling back on another day when I didn't happen to have `zeek-config` in my path, and I was and using `zkg install` to try to install the [zeek-community-id](https://github.com/corelight/zeek-community-id) package, but it failed with the error `Either 'zeek-config' must be in PATH or '--zeek-dist=<path>' used`. While I was able to force it to work by once again adding `zeek-config` to my path, the reference to `--zeek-dist` made me pause because the value that I'd have sent via that flag if I was calling the `configure` script directly is already present in `zkg config` (see the last line):

```
# zkg config
[sources]
zeek = https://github.com/zeek/packages
[paths]
state_dir = /root/.zkg
script_dir = /usr/local/zeek-3.1.4/share/zeek/site
plugin_dir = /usr/local/zeek-3.1.4/lib/zeek/plugins
zeek_dist = /root/zeek-3.1.4
```

As @ckreibich noted in the Slack thread:

> So, it's not a bug, but I think a very reasonable feature request: allow zkg's config knowledge to be available in the plugin's configure script. We can look into that.

**Comment by jsiwek** [comment]
Use of `zeek-config` I recall usually results in building against a Zeek install-tree, while `zeek_dist` / `--zeek-dist` is explicitly for building against a source-tree.  Don't think there's really any difference in end-result, but mentioning it because `zeek_dist` / `--zeek-dist` is kind of legacy thing that just predates the ability to build against an install-tree (and/or is rooted in a time when `zeek-config`/`bro-config` didn't even exist).  So adding logic to `zkg` to pass along the `zeek_dist` info to packages gives me pause since it introduces tighter coupling / specific usage of that legacy component.  Maybe the more generic enhancement to make is:

1. Have `zkg autoconfig` record the path to `zeek-config`:

```
# zkg config
...
[paths]
zeek_config = /usr/local/zeek-3.1.4/bin/zeek-config
...
```

2. If there's not a `zeek-config` in `PATH` at the time of installing/testing a package, but there is a `zeek_config` setting, then add its associated path to `PATH` before running the package's `build_command`/`test_command`.

i.e. move further along the path of `zeek-config` becoming the standardized path-reporting-tool.

**Comment by ckreibich** [comment]
Good thoughts Jon. I definitely agree that `zeek_dist` isn't really the way forward. All "modern" packages I've seen remove the `--bro-dist` / `--zeek-dist` from their configure invocations, to move to install-tree builds. My thinking was simply that since `zkg` already relies on `zeek-config` to populate settings during autoconfig, it would make sense to share more of that as it launches the `build_command`. So I agree with your thinking here.

Do you think it might make sense to change step 2 to work the other way around? I.e., whenever `zkg autoconfig` has recorded a path to `zeek-config`, use that first, and only fall back to the PATH when not present? I'm saying this only because I'd consider the other settings in `zkg`'s config file to be authoritative over other environment variables too. (For example, it doesn't matter if I have a `ZEEK_PLUGIN_PATH` set; `zkg`'s `plugin_dir` will still apply.

**Comment by jsiwek** [comment]
> Do you think it might make sense to change step 2 to work the other way around? I.e., whenever `zkg autoconfig` has recorded a path to `zeek-config`, use that first, and only fall back to the PATH when not present?

Yeah, thanks for that idea, I agree.

**Comment by grigorescu** [comment]
Hmm, I disagree. Generally, I expect the order of precedence to be:

1. Command line options I explicitly set,
2. Environment variables,
3. Local config file,
4. Global config file.

In this specific case, I expect to be able to override zkg behavior with an environment variable, regardless of the config file. Doing otherwise seems too "magical" and breaks the principle of least surprise.

Perhaps we just need a `ZEEK_CONFIG` env var, which points to the `zeek-config` that the user wishes to use. I think part off my hold-up is `zkg` both relying on and modifying `PATH` in this scenario.

If we're moving towards increasing the reliance on `zeek-config`, why even bother with `script_dir` and `plugin_dir`? I think that the goal should be to find the right `zeek-config`, and then everything else follows from that.

> I.e., whenever zkg autoconfig has recorded a path to zeek-config, use that first, and only fall back to the PATH when not present? I'm saying this only because I'd consider the other settings in zkg's config file to be authoritative over other environment variables too. (For example, it doesn't matter if I have a ZEEK_PLUGIN_PATH set; zkg's plugin_dir will still apply.

So, what would this look like in practice? Let's say that my autoconfig path to `zeek-config` is no longer valid, but my `plugin_dir` IS valid. Using the proposed method, `zkg` would check PATH and find the correct `zkg`, but still use the old `plugin_dir` and not provide the user a way to override that via env vars?

**Comment by jsiwek** [comment]
> Generally, I expect the order of precedence to be:
> 
> 1. Command line options I explicitly set,
> 2. Environment variables,
> 3. Local config file,
> 4. Global config file.

Yeah, generally, but would suggest working through any specific use-case we have in mind to see how they fit into that model in practice.

This is the order followed by the `zkg` config file itself that does match your expectation:

https://github.com/zeek/package-manager/blob/4fa8b3ed18131ab8f8a099fb44018833bdb428f2/zkg.config#L3-L9

And, since `ZEEK_PLUGIN_PATH` was specifically mentioned, a clarification on how `zkg` currently handles it: when running a `test_command` from a package, it does actually prefer grabbing that from the environment first:

https://github.com/zeek/package-manager/blob/4fa8b3ed18131ab8f8a099fb44018833bdb428f2/zeekpkg/manager.py#L2151

But let's also remember this env. var. is specific to `zeek`'s usage, not `zkg`'s, and don't think it makes sense to confound it with `zkg`'s choice of installation dirs, which is currently customized entirely through the config file, not any specific env. vars.

> In this specific case, I expect to be able to override zkg behavior with an environment variable, regardless of the config file. Doing otherwise seems too "magical" and breaks the principle of least surprise.

What's an example of a situation where you'd want to override a `zeek_config` that's been pre-set in the config file?  A hunch is that you probably won't want to use a different `zeek-config` between invocations of `zkg`.  You'd instead either use multiple `zkg` config files to manage more than one Zeek installation, or else find it simpler to update the config file with a new `zeek-config` path to use from then on.

I wonder if it's more magical/surprising for have a change to PATH between `zkg` invocations to suddenly be finding a different `zeek-config` and unintentionally trying to build/install things against the wrong Zeek installation.

> Perhaps we just need a `ZEEK_CONFIG` env var, which points to the `zeek-config` that the user wishes to use.

Probably wouldn't have an objection of teaching `zkg` to look for that env. var. if there is a use a case, but still wonder if temporarily changing the `zeek-config` to point to something else than what the config file uses is begging for inconsistent/broken installations.

> If we're moving towards increasing the reliance on `zeek-config`, why even bother with `script_dir` and `plugin_dir`? I think that the goal should be to find the right `zeek-config`, and then everything else follows from that.

True that we could change the `zkg` behavior to dynamically ask `zeek-config` for `script_dir` and `plugin_dir` on every invocation instead of have them recorded in the config file, but might also be an opportunity to warn the user about unexpected path mismatches.  E.g if they have a `zeek_config` configured, then invoke the associated `zeek-config` at startup and check if `script_dir` and `plugin_dir` still match what's in the config file.  If they don't match, warn/prompt the user that they'll likely need to update their config to avoid inconsistent/broken installation.

> > I.e., whenever zkg autoconfig has recorded a path to zeek-config, use that first, and only fall back to the PATH when not present? I'm saying this only because I'd consider the other settings in zkg's config file to be authoritative over other environment variables too. (For example, it doesn't matter if I have a ZEEK_PLUGIN_PATH set; zkg's plugin_dir will still apply.
> 
> So, what would this look like in practice? Let's say that I upgraded Zeek, and my autoconfig paths are no longer valid. Using the proposed method, `zkg` would check PATH and find the correct `zeek-config`, but still use the old `plugin_dir` and not provide the user a way to override that via env vars?

Here's my walk through of the "upgrading Zeek" scenario:

### Installing new Zeek to different prefix than old installation

If `zkg` had recorded the original path of `zeek-config` (and `script_dir` / `plugin_dir`) from the old installation, then any further `zkg` operations continue to use those same paths and operates on the old installation only.

If the user updates their `PATH` to use the new Zeek install-tree.  There's no inconsistencies as `zkg` continues using paths associated with the old Zeek install.  There's also the option of warning the user that `zkg` is using a `zeek-config` that differs from what's found in `PATH` and that they may want to update their `zkg` config -- running `zkg autoconfig` at that point would replace all the settings in the config file with what the new Zeek installation's `zeek-config` reports.

### Installing new Zeek to same prefix as old installation

Seems there's less that can go wrong here since all the paths should still agree, but if we had pre-recorded the `zeek-config` paths, we can do the additional sanity check I mentioned before of looking for consistent `script_dir` / `plugin_dir`.

In any case, prioritizing the `zkg` config file's `zeek-config` over what's in `PATH` seems to have only benefits of helping catch mistakes/inconsistencies while prioritizing `zeek-config` from `PATH` may lead to more of them?

**Comment by ckreibich** [comment]
Here's my take. I don't mean to violate the usual cascade of overrides, that's of course a good convention. But using `PATH` as the override mechanism, particularly a partial one, would be really confusing. When I use `zkg --configfile` — pretty explicit! — then it's counter-intuitive to have `PATH` override it. I don't think that the Zeek installation I'm running right now should automatically always be the one I operate on via `zkg`.

I personally like the idea of embracing additional environment variables. We haven't really talked about how `zkg` will pass its knowledge into the `build_command`. If we define a new namespace of environment variables — say `ZKG_*`, with `ZKG_ZEEK_CONFIG` as an example — then we could have things work as follows:

- New `zkg autoconfig` runs store the location of `zeek-config` in its config file
- `zkg` passes this value as `ZKG_ZEEK_CONFIG` into the environment of the `build_command`
- We expand the default `configure` template to honor this variable.

These three together address the problem Phil is trying to resolve. Build scripts continue to have all freedom and get an explicit signal from `zkg` about the settings it would like it to use.

For your override concerns, Vlad:

- If you have `ZKG_ZEEK_CONFIG` in your environment when you invoke `zkg`, that can override what's in the config file.
- For existing setups, nothing changes, you still need to have a `zeek-config` in your path to be able to install a plugin.

Regarding keeping `script_dir` and `plugin_dir` in the config file in addition to `zeek-config`, I think there's value in supporting that, for customized settings. I don't always want to have to stick to what `zeek-config` proposes. And, those setting too could be passed into the `build_command`.

**Comment by jsiwek** [comment]
> * New `zkg autoconfig` runs store the location of `zeek-config` in its config file
> * `zkg` passes this value as `ZKG_ZEEK_CONFIG` into the environment of the `build_command`
> * We expand the default `configure` template to honor this variable.

This way sounds fine to me, too, but it does require individual packages to update to the new `configure` template, so it will take time for it to become a generalized solution.  Also would suggest using `ZEEK_CONFIG` for the `configure` template itself as people running `./configure` directly and that know nothing of `zkg` may find it useful.

**Comment by ckreibich** [comment]
Yeah, I agree with those concerns Jon. I don't feel strongly here.

**Comment by grigorescu** [comment]
Thanks to you both for the detailed responses; good discussion. I definitely had some misunderstandings regarding how `zkg` handles `ZEEK_PLUGIN_PATH`.

> * New zkg autoconfig runs store the location of zeek-config in its config file
> * zkg passes this value as **ZEEK_CONFIG** into the environment of the build_command
> * We expand the default configure template to honor this variable.

Sounds good to me too. I do also like the idea of warning the user if what's being reported by `zeek-config` doesn't match the config.

Apologies for the scope creep, but one more edge case to consider -- what if there is no configuration at all? In this case, running `zkg install` will report that it installed correctly, but from what I can tell, it will only "install" into `~/.zkg/plugin_dir` and `~/.zkg/script_dir`. Would better behavior be that if `zkg` has not been configured, but `zeek-config` is in the path, to install to the directories that `zeek-config` returns?

**Comment by jsiwek** [comment]
> what if there is no configuration at all? In this case, running `zkg install` will report that it installed correctly, but from what I can tell, it will only "install" into `~/.zkg/plugin_dir` and `~/.zkg/script_dir`. Would better behavior be that if `zkg` has not been configured, but `zeek-config` is in the path, to install to the directories that `zeek-config` returns?

Yeah, kind of falls under the same idea as in #63.  I'd probably just generalize it to: if we're creating the state-dir, `~/.zkg`, for the first time and running any other command besides `zkg autoconfig`, we can warn the user that they might want to stop and run `zkg autoconfig` before proceeding.
